#**LECCION 1**

##EJERCICIO 1
El propósito de esta actividad es poner en práctica la creación y manipulación de matrices en NumPy, así como la aplicación de operaciones básicas y comprensión del comportamiento de la indexación.

Paso a paso:

1. Importa la librería NumPy como пр.

2. Crea una matriz A de 3x3 con números enteros del 1 al 9.

3. Realiza las siguientes operaciones: Suma 10 a cada elemento de la matriz.

Consigna:

Crea una matriz 3x3 utilizando np.array() a partir de listas de listas. Luego, aplica operaciones matemáticas básicas (suma, resta y multiplicación por un escalar), y realiza una selección condicional de elementos mayores a 5. Finalmente, analiza si el resultado de una asignación es una copia o una referencia.

Multiplica la matriz por 2.

Extrae los elementos mayores a 5 utilizando selección condicional.

4. Asigna la matriz a una nueva variable B = Ay modifica un valor de B. ¿Cambió también A? Ahora usá B = A.copy() y repeti la modificación. ¿Qué ocurrió con A esta vez?

##EJERCICIO  2
Contexto: H

Esta actividad busca fortalecer el dominio sobre las funciones predefinidas para crear arreglos, generar datos de prueba y reestructurar la forma de los mismos usando reshape().

Consigna:

Utiliza funciones de NumPy como np.arange(). np.linspace() y np.random.randint() para generar distintos tipos de arreglos. Luego redimensionalos con reshape() y aplica funciones matemáticas a los mismos.

Tiempo: 30 minutos

Paso a paso:

1. Crea un arreglo x con np.arange(1, 17) y redimensionalo en una matriz de 4x4.

2. Genera un arreglo de 10 elementos equidistantes entre 0 y 100 usando np.linspace().

3. Crea una matriz aleatoria de 3x3 con enteros entre 1 y 20 usando np.random.randint().

4. Aplica np.sqrt() y np.log() sobre las matrices creadas (cuando sea posible).

5. ¿Qué diferencias notas en la forma de aplicar funciones sobre arreglos enteros vs flotantes

In [ ]:
import numpy as np

# Paso 2: Crear matriz 3x3 del 1 al 9
A = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
])
print("Matriz A original:\n", A)

# Sumar 10 a cada elemento
A_suma10 = A + 10
print("\nMatriz A + 10:\n", A_suma10)

# Multiplicar por 2
A_mul2 = A * 2
print("\nMatriz A * 2:\n", A_mul2)

# Selección condicional: elementos > 5
mayores_5 = A[A > 5]
print("\nElementos de A mayores a 5:\n", mayores_5)

# Asignación por referencia
B = A
B[0, 0] = 99
print("\nA después de modificar B (por referencia):\n", A)

# Asignación por copia
C = A.copy()
C[0, 0] = 1  # Restauramos valor en la copia
print("\nC (copia de A con modificación):\n", C)
print("A después de modificar C (no cambia):\n", A)

Matriz A original:
 [[1 2 3]
 [4 5 6]
 [7 8 9]]

Matriz A + 10:
 [[11 12 13]
 [14 15 16]
 [17 18 19]]

Matriz A * 2:
 [[ 2  4  6]
 [ 8 10 12]
 [14 16 18]]

Elementos de A mayores a 5:
 [6 7 8 9]

A después de modificar B (por referencia):
 [[99  2  3]
 [ 4  5  6]
 [ 7  8  9]]

C (copia de A con modificación):
 [[1 2 3]
 [4 5 6]
 [7 8 9]]
A después de modificar C (no cambia):
 [[99  2  3]
 [ 4  5  6]
 [ 7  8  9]]


In [ ]:
# 1. Arreglo del 1 al 16 y reshape
x = np.arange(1, 17)
x_matrix = x.reshape(4, 4)
print("Matriz 4x4 con np.arange:\n", x_matrix)

# 2. Arreglo de 10 valores equidistantes entre 0 y 100
lin = np.linspace(0, 100, 10)
print("\nArreglo con np.linspace:\n", lin)

# 3. Matriz aleatoria de 3x3 entre 1 y 20
random_mat = np.random.randint(1, 21, (3, 3))
print("\nMatriz aleatoria 3x3:\n", random_mat)

# 4. Funciones matemáticas
print("\nRaíz cuadrada (sqrt) de linspace:\n", np.sqrt(lin))
print("\nLogaritmo natural (log) de linspace + 1:\n", np.log(lin + 1))  # +1 para evitar log(0)

print("\nRaíz cuadrada de matriz aleatoria:\n", np.sqrt(random_mat))
print("\nLogaritmo de matriz aleatoria:\n", np.log(random_mat))

Matriz 4x4 con np.arange:
 [[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]
 [13 14 15 16]]

Arreglo con np.linspace:
 [  0.          11.11111111  22.22222222  33.33333333  44.44444444
  55.55555556  66.66666667  77.77777778  88.88888889 100.        ]

Matriz aleatoria 3x3:
 [[ 7 13  9]
 [16  5  1]
 [ 4 17 20]]

Raíz cuadrada (sqrt) de linspace:
 [ 0.          3.33333333  4.71404521  5.77350269  6.66666667  7.45355992
  8.16496581  8.81917104  9.42809042 10.        ]

Logaritmo natural (log) de linspace + 1:
 [0.         2.4941233  3.14510967 3.5361167  3.81649058 4.03522344
 4.21459369 4.36663095 4.49857434 4.61512052]

Raíz cuadrada de matriz aleatoria:
 [[2.64575131 3.60555128 3.        ]
 [4.         2.23606798 1.        ]
 [2.         4.12310563 4.47213595]]

Logaritmo de matriz aleatoria:
 [[1.94591015 2.56494936 2.19722458]
 [2.77258872 1.60943791 0.        ]
 [1.38629436 2.83321334 2.99573227]]


#LECCIÓN 2

##Ejercicio 1
Primer acercamiento al análisis exploratorio con Pandas

Contexto: H

El objetivo es familiarizarse con la estructura de un DataFrame y aplicar los métodos básicos de exploración y resumen de datos para obtener una primera comprensión de su contenido.

Paso a paso:

Importa la librería pandas como pd.

Lee el archivo data.csv con pd.read_csv('data.csv') y asignalo a una variable llamada df.

Usa los métodos:

df.head() para ver los primeros registros.

Consigna:

Carga un archivo CSV con Pandas y explora su contenido utilizando métodos como head(), info() y describe(). Luego, extrae estadísticas clave de las columnas numéricas y verifica si hay valores nulos.

df.info() para revisar estructura y tipos de datos.

df.describe() para ver estadísticas básicas.

Aplica df.isnull().sum() para detectar valores nulos. Usa df.mean(), df.median() y df.std() para obtener medidas adicionales.

Tiempo: 20 minutos

Comparte una breve interpretación de lo que observaste.



##Ejercicio  2

Paso a paso:

1. Abre tu entorno de trabajo preferido (VS Code, Jupyter, Colab, etc.).

2. Crea un nuevo archivo o celda de código.

3. Escribe un programa que:

Solicite al usuario su nombre y edad.

Use una estructura condicional (if) para verificar si es mayor o menor de edad.

Imprima un mensaje de saludo que incluya el nombre y la condición (mayor o menor).

4. Probalo con diferentes edades para verificar que funcione correctamente.

5. (Opcional) Agrega una función llamada es_mayor() que reciba la edad y retorne True o False.

In [1]:
import pandas as pd
import numpy as np

# 1. Crear datos simulados
np.random.seed(0)

n = 100

data = {
    "Edad": np.random.randint(18, 65, n),
    "Salario": np.round(np.random.normal(35000, 8000, n), 2),
    "Compras_realizadas": np.random.randint(0, 30, n),
    "Departamento": np.random.choice(["Ventas", "Marketing", "TI", "Finanzas"], n),
    "Suscripto_boletin": np.random.choice([True, False], n)
}

df_simulado = pd.DataFrame(data)

# 2. Introducir valores nulos en la columna 'Salario'
indices_nulos = np.random.choice(df_simulado.index, size=5, replace=False)
df_simulado.loc[indices_nulos, "Salario"] = np.nan

# 3. Guardar como CSV
df_simulado.to_csv("data.csv", index=False)
print("Archivo 'data.csv' creado exitosamente.")


import pandas as pd

# 1. Cargar el archivo CSV
df = pd.read_csv("data.csv")

# 2. Primeras filas
print("Primeras filas del DataFrame:\n")
print(df.head())

# 3. Información general
print("\nInformación del DataFrame:\n")
print(df.info())

# 4. Estadísticas básicas
print("\nResumen estadístico de columnas numéricas:\n")
print(df.describe())

# 5. Verificación de valores nulos
print("\nValores nulos por columna:\n")
print(df.isnull().sum())

# 6. Medidas adicionales
print("\nMedia de columnas numéricas:")
print(df.mean(numeric_only=True))

print("\nMediana de columnas numéricas:")
print(df.median(numeric_only=True))

print("\nDesviación estándar de columnas numéricas:")
print(df.std(numeric_only=True))


Archivo 'data.csv' creado exitosamente.
Primeras filas del DataFrame:

   Edad   Salario  Compras_realizadas Departamento  Suscripto_boletin
0    62  35214.95                  29           TI               True
1    18  43446.70                   2    Marketing               True
2    21  34414.53                  15           TI               True
3    21  29674.23                  27     Finanzas               True
4    57  34647.10                  21           TI              False

Información del DataFrame:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Edad                100 non-null    int64  
 1   Salario             95 non-null     float64
 2   Compras_realizadas  100 non-null    int64  
 3   Departamento        100 non-null    object 
 4   Suscripto_boletin   100 non-null    bool   
dtypes: bool(1), float64(1), int64

El DataFrame contiene 100 registros y 5 columnas: tres numéricas (Edad, Salario, Compras_realizadas), una categórica (Departamento) y una booleana (Suscripto_boletin).

El método info() muestra que hay valores nulos en la columna Salario, lo cual es importante considerar antes de aplicar modelos o gráficos.

describe() revela el rango, media, y dispersión de las variables numéricas. Por ejemplo:

La edad media ronda los 41 años.

Los salarios tienen una media cercana a los $35,000, con cierta dispersión.

Las compras oscilan entre 0 y 29.

Las funciones mean(), median() y std() ayudan a analizar la simetría y dispersión de los datos.

In [ ]:
##EJERCICIO 2
# Función opcional
def es_mayor(edad):
    return edad >= 18

# Solicitar datos al usuario
nombre = input("¿Cuál es tu nombre? ")
edad = int(input("¿Cuántos años tienes? "))

# Verificar si es mayor de edad
if es_mayor(edad):
    print(f"Hola {nombre}, eres mayor de edad.")
else:
    print(f"Hola {nombre}, eres menor de edad.")

¿Cuál es tu nombre? Lo
¿Cuántos años tienes? 32
Hola Lo, eres mayor de edad.


#**Lección 3**

los ejercicio estan en otro notebook

#LECCION 4
CARGA Y EXPORTACION DE ARCHIVOS CSV
Ejercicio  1
Contexto: H

En esta actividad, los estudiantes aprenderán a importar datos desde archivos .csv, adaptarlos a las necesidades del análisis, y luego exportarlos nuevamente con las configuraciones adecuadas.

Consigna:

Utiliza Pandas para cargar un archivo CSV sin encabezado, renombrar las columnas, limpia los datos nulos y exporta el resultado a un nuevo archivo.csv con delimitador personalizado.

Tiempo: 30 minutos

X

32

> :

Paso a paso:

1. Abre tu entorno de trabajo preferido (VS Code, Jupyter, Colab, etc.).

2. Importa Pandas (import pandas as pd).

3. Lee el archivo datos.csv usando header=None y asigna nombres personalizados con names=['nombre', 'edad', 'ciudad'].

4. Reemplaza valores nulos o símbolos como '?' usando na_values y fillna().

5. Elimina filas duplicadas si las hubiera con drop_duplicates().

6. Exporta el DataFrame limpio con to_csv(), usando sep=';' y encoding='utf-8'.

##Ejercicio 2
LECTURA DE DATOS DESDE EXCEL Y EXTRACCION WEB
Contexto:+H

Comprender cómo obtener datos desde archivos Excel y páginas web, y cómo unificarlos en un solo DataFrame para análisis conjunto.

Consigna:

Lee datos desde un archivo .xlsx y desde una tabla HTML de una web. Combina ambos conjuntos en un solo DataFrame y exporta el resultado como archivo Excel con múltiples hojas.

Paso a paso:

1. Asegurate de tener instalada la librería openpyxl.

2. Carga un archivo Excel con read_excel('archivo.xlsx', sheet_name='Ventas').

3. Usa read_html() para extraer la primera tabla de una web (por ejemplo, una página de estadísticas).

4. Unifica ambos conjuntos con concat() o merge() según el caso.

5. Exporta el resultado a un archivo resumen.xlsx con to_excel(), incluyendo ambos DataFrames en hojas diferentes usando Excel Writer.

Tiempo : 30 minutos

In [2]:
#EJERCICIO 1

import pandas as pd

# 1. Simular datos con errores y sin encabezado
datos = [
    ["Ana", 28, "Santiago"],
    ["Luis", "?", "Valparaíso"],
    ["Camila", 35, "Santiago"],
    ["Pedro", "", "Antofagasta"],
    ["Ana", 28, "Santiago"],  # Fila duplicada
    ["Jorge", 22, "?"]
]

# Guardar como CSV sin encabezado
df_sin_encabezado = pd.DataFrame(datos)
df_sin_encabezado.to_csv("datos.csv", index=False, header=False)

# 2. Cargar archivo CSV sin encabezado y asignar nombres de columna
df = pd.read_csv("datos.csv", header=None, names=['nombre', 'edad', 'ciudad'], na_values=['?', ''])

# 3. Mostrar DataFrame original
print("DataFrame original con posibles valores nulos o inválidos:")
print(df)

# 4. Reemplazar valores nulos
df['edad'] = pd.to_numeric(df['edad'], errors='coerce')  # Asegura que 'edad' sea numérica
df['edad'] = df['edad'].fillna(df['edad'].median())
df['ciudad'] = df['ciudad'].fillna("Desconocido")

# 5. Eliminar filas duplicadas
df_limpio = df.drop_duplicates()

# 6. Mostrar DataFrame limpio
print("\nDataFrame limpio:")
print(df_limpio)

# 7. Exportar DataFrame limpio con separador ';' y codificación utf-8
df_limpio.to_csv("datos_limpios.csv", index=False, sep=';', encoding='utf-8')
print("\nArchivo 'datos_limpios.csv' exportado correctamente.")


DataFrame original con posibles valores nulos o inválidos:
   nombre  edad       ciudad
0     Ana  28.0     Santiago
1    Luis   NaN   Valparaíso
2  Camila  35.0     Santiago
3   Pedro   NaN  Antofagasta
4     Ana  28.0     Santiago
5   Jorge  22.0          NaN

DataFrame limpio:
   nombre  edad       ciudad
0     Ana  28.0     Santiago
1    Luis  28.0   Valparaíso
2  Camila  35.0     Santiago
3   Pedro  28.0  Antofagasta
5   Jorge  22.0  Desconocido

Archivo 'datos_limpios.csv' exportado correctamente.


In [4]:
#EJERCICIO 2

import pandas as pd
import os

# Instalar librerías necesarias
!pip install -q kagglehub openpyxl

import kagglehub

# === 1.b. Cargar archivo Excel: Liga Española ===
print("\n=== 1.b. Cargar archivo Excel: Liga Española ===")

# Descargar el archivo Excel desde Kaggle (requiere estar autenticado con kaggle.json)
path_excel = kagglehub.dataset_download("jbatistazerpa/liga-espaola-2023-2024-resultados")
excel_files = os.listdir(path_excel)
print("Archivos disponibles:", excel_files)

# Cargar el archivo Excel
df_excel = pd.read_excel(os.path.join(path_excel, "LigaEspanola2023-2024-Resultados.xlsx"))
print("\n▶ Primeras 5 filas del archivo Excel:")
display(df_excel.head())

# === 1.c. Cargar tabla web: Wikipedia Población Mundial ===
print("\n=== 1.c. Cargar tabla web: Wikipedia Población Mundial ===")

# Leer tabla desde Wikipedia
url_web = "https://en.wikipedia.org/wiki/List_of_countries_and_dependencies_by_population"
tablas = pd.read_html(url_web)

# Mostramos cuántas tablas se encontraron
print(f"▶ Se encontraron {len(tablas)} tablas.")

# Elegimos la primera tabla (generalmente es la principal)
df_web = tablas[0]
print("\n▶ Primeras 5 filas de la tabla web:")
display(df_web.head())

# === 2. Unificación de datos (opcional, sólo si queremos combinarlos en una hoja extra) ===
# Para simplificar la unión, tomamos solo algunas columnas comunes
df_excel_short = df_excel.iloc[:, :3]  # Primeras 3 columnas
df_web_short = df_web.iloc[:, :3]     # Primeras 3 columnas de la tabla web
df_unido = pd.concat([df_excel_short, df_web_short], ignore_index=True)

# === 3. Exportar a resumen.xlsx con múltiples hojas ===
print("\n=== 3. Exportar a resumen.xlsx ===")
with pd.ExcelWriter("resumen.xlsx", engine="openpyxl") as writer:
    df_excel.to_excel(writer, sheet_name="Liga Española", index=False)
    df_web.to_excel(writer, sheet_name="Población", index=False)
    df_unido.to_excel(writer, sheet_name="Unido", index=False)

print("Archivo 'resumen.xlsx' exportado correctamente con 3 hojas.")




=== 1.b. Cargar archivo Excel: Liga Española ===


100%|██████████| 14.5k/14.5k [00:00<00:00, 15.7MB/s]

Extracting files...
Archivos disponibles: ['LigaEspanola2023-2024-Resultados.xlsx']

▶ Primeras 5 filas del archivo Excel:


Round       Date   Home Team       Away Team  Home Score  Away Score  \
0     33 2024-04-29   Barcelona        Valencia           4           2   
1     33 2024-04-28       Betis         Sevilla           1           1   
2     33 2024-04-28  Villarreal  Rayo Vallecano           3           0   
3     33 2024-04-28  Granada CF         Osasuna           3           0   
4     33 2024-04-28    Cadiz CF        Mallorca           1           1   

   Total Goals Result  
0            6   Home  
1            2    Tie  
2            3   Home  
3            3   Home  
4            2    Tie


=== 1.c. Cargar tabla web: Wikipedia Población Mundial ===
▶ Se encontraron 3 tablas.

▶ Primeras 5 filas de la tabla web:


Location  Population % of world         Date  \
0          World  8232000000       100%  13 Jun 2025   
1          India  1413324000      17.3%   1 Mar 2025   
2          China  1408280000      17.2%  31 Dec 2024   
3  United States   340110988       4.2%   1 Jul 2024   
4      Indonesia   282477584       3.5%  30 Jun 2024   

  Source (official or from the United Nations) Notes  
0                          UN projection[1][3]   NaN  
1                       Official projection[4]   [b]  
2                         Official estimate[5]   [c]  
3                         Official estimate[6]   [d]  
4                National annual projection[7]   NaN


=== 3. Exportar a resumen.xlsx ===
Archivo 'resumen.xlsx' exportado correctamente con 3 hojas.


#LECCION 5

##Ejercicio  1
LIMPIEZA DE DATOS EN UN ECOMMERCE
Contexto:

Una tienda online desea analizar las ventas del último año, pero los datos vienen de múltiples fuentes con errores comunes: valores nulos, duplicados y formatos inconsistentes.

Paso a paso:

1. Carga el dataset con pd.read_csv().

2. Detecta y elimina filas duplicadas.

Consigna:

3. Identifica columnas con valores nulos y decidísi deben eliminarse o imputarse.

Usa Pandas para limpiar el archivo ventas_ecommerce.csv, aplicando técnicas de detección de duplicados, reemplazo de valores erróneos y normalización de formatos.

4. Reemplaza valores incorrectos en la columna Método de pago (ej: "tarjeta", "efetivo") usando replace().

Tiempo : 30 minutos

5. Asegurate de que las fechas estén en formato datetime y los precios en float.


##Ejercicio  2
TRASNSFOMACION Y ENRIQUECIIMENTO DE DATOS DE RRHH
Contexto: +

Un área de Recursos Humanos necesita preparar un 1. informe de rotación laboral con datos de empleados. Se requiere agrupar, discretizar y enriquecer la información.

Consigna:

Aplica transformaciones avanzadas a un dataset de empleados (empleados.csv) para crear nuevas variables y preparar los datos para análisis.

Tiempo: 30 minutos

Paso a paso:

Agrupa los empleados por área y calcula el promedio de antigüedad y edad.

2. Discretiza la edad en categorías: Joven (≤30). Medio (31-45), Senior (46+).

3.

Crea una nueva columna con apply() para clasificar empleados según su permanencia (Antigüedad > 5 años = "Estable").

4. Exporta el nuevo DataFrame limpio Con to_csv().

In [6]:
#EJERCICIO 1:

import pandas as pd

# 1. Simular archivo ventas_ecommerce.csv con errores comunes
datos = {
    "Fecha": ["2023-01-15", "2023-01-15", "2023/02/01", "", "2023-03-05", "2023-03-05"],
    "Cliente": ["Ana", "Ana", "Luis", "Camila", "Pedro", "Pedro"],
    "Producto": ["Laptop", "Laptop", "Mouse", "Teclado", "Monitor", "Monitor"],
    "Precio": ["1000", "1000", "25.5", "30.0", "500,00", "500,00"],
    "Método de pago": ["tarjeta", "tarjeta", "efetivo", "efectivo", "Tarjeta", "Tarjeta"]
}

# Guardamos como archivo CSV con errores
df_simulado = pd.DataFrame(datos)
df_simulado.to_csv("ventas_ecommerce.csv", index=False)

# 2. Cargar el archivo
df = pd.read_csv("ventas_ecommerce.csv")
print("=== Datos originales ===")
print(df)

# 3. Eliminar duplicados
df = df.drop_duplicates()
print("\n=== Sin duplicados ===")
print(df)

# 4. Identificar valores nulos
print("\n=== Valores nulos por columna ===")
print(df.isnull().sum())

# Eliminar filas con fecha vacía (podrías imputar si prefieres)
df = df[df["Fecha"].notnull()]

# 5. Normalizar la columna de fechas
df["Fecha"] = pd.to_datetime(df["Fecha"], errors="coerce")

# 6. Convertir precios a float (corrigiendo comas)
df["Precio"] = df["Precio"].str.replace(",", ".", regex=False)
df["Precio"] = pd.to_numeric(df["Precio"], errors="coerce")

# 7. Corregir errores en método de pago
df["Método de pago"] = df["Método de pago"].str.lower().replace({
    "efetivo": "efectivo",
    "tarjeta": "tarjeta",
    "Tarjeta": "tarjeta"
})

# 8. Resultado limpio
print("\n=== Datos limpios ===")
print(df)

# 9. Exportar archivo limpio
df.to_csv("ventas_ecommerce_limpio.csv", index=False)
print("\nArchivo 'ventas_ecommerce_limpio.csv' exportado correctamente.")


=== Datos originales ===
        Fecha Cliente Producto  Precio Método de pago
0  2023-01-15     Ana   Laptop    1000        tarjeta
1  2023-01-15     Ana   Laptop    1000        tarjeta
2  2023/02/01    Luis    Mouse    25.5        efetivo
3         NaN  Camila  Teclado    30.0       efectivo
4  2023-03-05   Pedro  Monitor  500,00        Tarjeta
5  2023-03-05   Pedro  Monitor  500,00        Tarjeta

=== Sin duplicados ===
        Fecha Cliente Producto  Precio Método de pago
0  2023-01-15     Ana   Laptop    1000        tarjeta
2  2023/02/01    Luis    Mouse    25.5        efetivo
3         NaN  Camila  Teclado    30.0       efectivo
4  2023-03-05   Pedro  Monitor  500,00        Tarjeta

=== Valores nulos por columna ===
Fecha             1
Cliente           0
Producto          0
Precio            0
Método de pago    0
dtype: int64

=== Datos limpios ===
       Fecha Cliente Producto  Precio Método de pago
0 2023-01-15     Ana   Laptop  1000.0        tarjeta
2        NaT    Luis    Mo

In [7]:
#EJERCICIO 2:

import pandas as pd

# 1. Simular datos de empleados
data = {
    "ID": [1, 2, 3, 4, 5, 6, 7],
    "Nombre": ["Ana", "Luis", "María", "Pedro", "Lucía", "Carlos", "Elena"],
    "Edad": [25, 38, 29, 46, 33, 52, 41],
    "Antigüedad": [2, 7, 4, 10, 5, 12, 6],
    "Área": ["Ventas", "TI", "Ventas", "TI", "RRHH", "RRHH", "Ventas"]
}

# Crear DataFrame y guardar como CSV
df_empleados = pd.DataFrame(data)
df_empleados.to_csv("empleados.csv", index=False)

# 2. Cargar el archivo
df = pd.read_csv("empleados.csv")
print("=== Datos originales ===")
print(df)

# 3. Agrupar por área: promedio de antigüedad y edad
agrupado = df.groupby("Área")[["Edad", "Antigüedad"]].mean().round(1)
print("\n=== Promedio de edad y antigüedad por área ===")
print(agrupado)

# 4. Discretizar edad en categorías
def clasificar_edad(edad):
    if edad <= 30:
        return "Joven"
    elif edad <= 45:
        return "Medio"
    else:
        return "Senior"

df["Grupo Edad"] = df["Edad"].apply(clasificar_edad)

# 5. Clasificar según permanencia
df["Clasificación Antigüedad"] = df["Antigüedad"].apply(lambda x: "Estable" if x > 5 else "Nuevo")

# 6. Mostrar DataFrame enriquecido
print("\n=== DataFrame enriquecido ===")
print(df)

# 7. Exportar el resultado limpio
df.to_csv("empleados_enriquecido.csv", index=False)
print("\nArchivo 'empleados_enriquecido.csv' exportado correctamente.")


=== Datos originales ===
   ID  Nombre  Edad  Antigüedad    Área
0   1     Ana    25           2  Ventas
1   2    Luis    38           7      TI
2   3   María    29           4  Ventas
3   4   Pedro    46          10      TI
4   5   Lucía    33           5    RRHH
5   6  Carlos    52          12    RRHH
6   7   Elena    41           6  Ventas

=== Promedio de edad y antigüedad por área ===
        Edad  Antigüedad
Área                    
RRHH    42.5         8.5
TI      42.0         8.5
Ventas  31.7         4.0

=== DataFrame enriquecido ===
   ID  Nombre  Edad  Antigüedad    Área Grupo Edad Clasificación Antigüedad
0   1     Ana    25           2  Ventas      Joven                    Nuevo
1   2    Luis    38           7      TI      Medio                  Estable
2   3   María    29           4  Ventas      Joven                    Nuevo
3   4   Pedro    46          10      TI     Senior                  Estable
4   5   Lucía    33           5    RRHH      Medio                    N

#LECCION 6
##Ejercicio 1
Análisis de ventas por categoría

Contexto:H

Una empresa de retail necesita analizar sus ventas segmentadas por categoría de producto y mes, para preparar un informe de performance.

Consigna:

Utiliza técnicas de agrupamiento y pivotado con Pandas para construir un resumen tabular de ventas mensuales por categoría.

Tiempo: 20 minutos

Paso a paso:

1. Carga el archivo ventas_productos.csv (contiene columnas como fecha, categoría, producto, ventas).

2. Convertir la columna fecha a tipo datetime y extraer el mes en una nueva columna.

3. Agrupa por categoría y mes usando groupby() y calculá la suma de ventas.

4. Pivota el resultado para que las categorías estén en filas y los meses en columnas.

5. Exporta el DataFrame pivotado como reporte_categorias.csv.

##Ejercicio  2
Integración de datos de usuarios y compras

Contexto: H

El equipo de marketing quiere unir datos de usuarios registrados con sus compras, para identificar patrones de comportamiento.

Paso a paso:

1. Carga dos archivos: usuarios.csv (con columnas user_id, segmento, registro_fecha) y compras.csv (con user_id, compra_id, monto).

Consigna:

Aplica merge() y groupby() en Pandas para analizar cuántas compras hizo cada tipo de usuario.

Tiempo: 20 minutos

2. Unir ambos datasets utilizando merge() por user_id.

3. Agrupar por segmento y contar la cantidad de compras realizadas.

4. Calcula también el monto total gastado por segmento usando .agg().

5. Mostrar los resultados ordenados de mayor a menor cantidad de compras.

In [8]:
# EJERCICIO 1: Análisis de ventas por categoría
import pandas as pd
from io import StringIO
from google.colab import files

# Simulamos archivo ventas_productos.csv
data = """
fecha,categoria,producto,ventas
2023-01-15,Tecnología,Laptop,1200
2023-01-20,Hogar,Aspiradora,300
2023-02-05,Tecnología,Smartphone,800
2023-02-15,Hogar,Licuadora,150
2023-03-01,Tecnología,Tablet,600
2023-03-05,Hogar,Refrigerador,900
2023-03-15,Deportes,Bicicleta,700
"""

# Cargar datos
df = pd.read_csv(StringIO(data))

display(df)

# 2. Convertir 'fecha' a datetime y extraer el mes
df['fecha'] = pd.to_datetime(df['fecha'])
df['mes'] = df['fecha'].dt.month

# 3. Agrupar por categoría y mes
resumen = df.groupby(['categoria', 'mes'])['ventas'].sum().reset_index()

display(resumen)



# 4. Pivotar el resultado
tabla_pivot = resumen.pivot(index='categoria', columns='mes', values='ventas').fillna(0)


# Mostrar resumen
print("=== Ventas por categoría y mes ===")
print(tabla_pivot)

# 5. Exportar a CSV
tabla_pivot.to_csv("reporte_categorias.csv", sep=";", encoding="utf-8")
files.download("reporte_categorias.csv")

fecha   categoria      producto  ventas
0  2023-01-15  Tecnología        Laptop    1200
1  2023-01-20       Hogar    Aspiradora     300
2  2023-02-05  Tecnología    Smartphone     800
3  2023-02-15       Hogar     Licuadora     150
4  2023-03-01  Tecnología        Tablet     600
5  2023-03-05       Hogar  Refrigerador     900
6  2023-03-15    Deportes     Bicicleta     700

categoria  mes  ventas
0    Deportes    3     700
1       Hogar    1     300
2       Hogar    2     150
3       Hogar    3     900
4  Tecnología    1    1200
5  Tecnología    2     800
6  Tecnología    3     600

=== Ventas por categoría y mes ===
mes              1      2      3
categoria                       
Deportes       0.0    0.0  700.0
Hogar        300.0  150.0  900.0
Tecnología  1200.0  800.0  600.0


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [9]:
# EJERCICIO 2: Integración de datos de usuarios y compras
import pandas as pd
from io import StringIO
from google.colab import files

# Simulamos usuarios.csv
usuarios_data = """
user_id,segmento,registro_fecha
1,Premium,2022-01-15
2,Básico,2022-02-20
3,Premium,2022-03-01
4,Básico,2022-04-10
"""

# Simulamos compras.csv
compras_data = """
user_id,compra_id,monto
1,101,250
1,102,400
2,103,150
3,104,300
3,105,100
3,106,200
4,107,50
"""

# Cargar los datasets
usuarios = pd.read_csv(StringIO(usuarios_data))
compras = pd.read_csv(StringIO(compras_data))

# 2. Unir por user_id
df_merged = pd.merge(compras, usuarios, on='user_id', how='left')

# 3. Agrupar por segmento
resumen_segmento = df_merged.groupby('segmento').agg(
    compras_realizadas=('compra_id', 'count'),
    monto_total=('monto', 'sum')
).reset_index()

# 4. Ordenar por cantidad de compras
resumen_segmento = resumen_segmento.sort_values(by='compras_realizadas', ascending=False)

# Mostrar resultado
print("=== Compras por segmento ===")
print(resumen_segmento)

# 5. Exportar a CSV
resumen_segmento.to_csv("reporte_segmentos.csv", sep=";", encoding="utf-8", index=False)
files.download("reporte_segmentos.csv")

=== Compras por segmento ===
  segmento  compras_realizadas  monto_total
1  Premium                   5         1250
0   Básico                   2          200


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>



## **Informe de Actividades - Análisis con Pandas**

### **Ejercicio 1: Análisis de ventas por categoría y mes**

**Contexto:**
Una empresa de retail desea analizar sus ventas mensuales por categoría de producto para generar un informe de performance. Se requiere transformar los datos para facilitar el análisis por mes y categoría.

---

### **Pasos realizados:**

**1. Carga de datos**

* Se utilizó un archivo simulado llamado `ventas_productos.csv`, el cual contiene columnas: `fecha`, `categoría`, `producto` y `ventas`.
* Se leyó el archivo con `pd.read_csv()`.

**2. Conversión de fechas**

* Se convirtió la columna `fecha` al tipo `datetime` con `pd.to_datetime()`.
* Luego se extrajo el número de mes (1 a 12) en una nueva columna `mes` utilizando `.dt.month`.

**3. Agrupación de datos**

* Se agruparon los datos por `categoría` y `mes`, calculando la suma de las ventas con `groupby()` y `sum()`.

**4. Pivotado**

* Se transformó la tabla agrupada en una estructura de resumen donde:

  * Las **filas** representan las **categorías**.
  * Las **columnas** representan los **meses**.
  * Las **celdas** contienen el **total de ventas** en ese mes para cada categoría.
* Esto se logró con `.pivot()`.

**5. Exportación**

* La tabla final fue exportada como `reporte_categorias.csv`, utilizando `to_csv()`, delimitador `;` y codificación `utf-8`.

---

### **Resultado esperado:**

Una tabla como esta:

| categoria  | 1    | 2   | 3   |
| ---------- | ---- | --- | --- |
| Deportes   | 0    | 0   | 700 |
| Hogar      | 300  | 150 | 900 |
| Tecnología | 1200 | 800 | 600 |

---

## **Ejercicio 2: Integración de datos de usuarios y compras**

**Contexto:**
El equipo de marketing desea unir datos de usuarios registrados con sus compras para identificar patrones según el tipo de usuario (segmento).

---

### **Pasos realizados:**

**1. Carga de dos datasets**

* Se utilizaron archivos simulados:

  * `usuarios.csv`: contiene `user_id`, `segmento`, `registro_fecha`.
  * `compras.csv`: contiene `user_id`, `compra_id`, `monto`.

**2. Unión de datos**

* Se unieron ambos datasets por la columna `user_id` usando `pd.merge()`.
* El resultado es una tabla con compras asociadas a cada usuario y su segmento.

**3. Agrupación y resumen**

* Se agrupó por `segmento` y se aplicaron dos operaciones:

  * Contar la cantidad de compras con `.count()`.
  * Calcular el total gastado con `.sum()` en la columna `monto`.

**4. Ordenar**

* Se ordenó el resumen final de mayor a menor según la cantidad de compras, usando `.sort_values()`.

**5. Exportación**

* Se exportó el resultado como `reporte_segmentos.csv`.

---

### **Resultado esperado:**

| segmento | compras\_realizadas | monto\_total |
| -------- | ------------------- | ------------ |
| Premium  | 5                   | 1250         |
| Básico   | 2                   | 200          |

---

##  Conclusión

Ambos ejercicios utilizaron funcionalidades fundamentales de **Pandas**:

* Limpieza y transformación de datos (`groupby`, `pivot`, `merge`)
* Conversión de tipos (`datetime`)
* Exportación a archivos reutilizables (`.csv`)

Estos procesos son altamente aplicables para generar **informes automáticos**, **dashboards** o **análisis exploratorios** en cualquier sector que maneje datos tabulados (retail, RRHH, marketing, salud, educación, etc.).



# 1. Arreglo del 1 al 16 y reshape
x = np.arange(1, 17)
x_matrix = x.reshape(4, 4)
print("Matriz 4x4 con np.arange:\n", x_matrix)

# 2. Arreglo de 10 valores equidistantes entre 0 y 100
lin = np.linspace(0, 100, 10)
print("\nArreglo con np.linspace:\n", lin)

# 3. Matriz aleatoria de 3x3 entre 1 y 20
random_mat = np.random.randint(1, 21, (3, 3))
print("\nMatriz aleatoria 3x3:\n", random_mat)

# 4. Funciones matemáticas
print("\nRaíz cuadrada (sqrt) de linspace:\n", np.sqrt(lin))
print("\nLogaritmo natural (log) de linspace + 1:\n", np.log(lin + 1))  # +1 para evitar log(0)

print("\nRaíz cuadrada de matriz aleatoria:\n", np.sqrt(random_mat))
print("\nLogaritmo de matriz aleatoria:\n", np.log(random_mat))

In [ ]:
import numpy as np

# 1. Crear arreglo del 1 al 16 y redimensionar en matriz 4x4
x = np.arange(1, 17)
x_matrix = x.reshape(4, 4)
print("Matriz 4x4 con np.arange:\n", x_matrix)

# 2. Crear 10 valores equidistantes entre 0 y 100
lin = np.linspace(0, 100, 10)
print("\nArreglo con np.linspace:\n", lin)

# 3. Matriz aleatoria de 3x3 entre 1 y 20
random_mat = np.random.randint(1, 21, (3, 3))
print("\nMatriz aleatoria 3x3:\n", random_mat)

# 4. Aplicar funciones matemáticas
print("\nRaíz cuadrada de linspace:\n", np.sqrt(lin))
print("\nLogaritmo natural de linspace + 1 (evitando log(0)):\n", np.log(lin + 1))

print("\nRaíz cuadrada de matriz aleatoria:\n", np.sqrt(random_mat))
print("\nLogaritmo natural de matriz aleatoria:\n", np.log(random_mat))

Matriz 4x4 con np.arange:
 [[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]
 [13 14 15 16]]

Arreglo con np.linspace:
 [  0.          11.11111111  22.22222222  33.33333333  44.44444444
  55.55555556  66.66666667  77.77777778  88.88888889 100.        ]

Matriz aleatoria 3x3:
 [[ 1 20 17]
 [ 6 19 20]
 [18 20 12]]

Raíz cuadrada de linspace:
 [ 0.          3.33333333  4.71404521  5.77350269  6.66666667  7.45355992
  8.16496581  8.81917104  9.42809042 10.        ]

Logaritmo natural de linspace + 1 (evitando log(0)):
 [0.         2.4941233  3.14510967 3.5361167  3.81649058 4.03522344
 4.21459369 4.36663095 4.49857434 4.61512052]

Raíz cuadrada de matriz aleatoria:
 [[1.         4.47213595 4.12310563]
 [2.44948974 4.35889894 4.47213595]
 [4.24264069 4.47213595 3.46410162]]

Logaritmo natural de matriz aleatoria:
 [[0.         2.99573227 2.83321334]
 [1.79175947 2.94443898 2.99573227]
 [2.89037176 2.99573227 2.48490665]]
